File: code/dart/test/http/http_client_test.dart
import 'dart:convert';

import 'package:http/http.dart' as http;
import 'package:http/testing.dart';
import 'package:service_client/service_client.dart';
import 'package:test/test.dart';

void main() {
  group('httpClient()', () {
    test('returns parsed data on success', () async {
      final mock = MockClient((_) async {
        return http.Response(
          jsonEncode({'id': 1, 'title': 'Test'}),
          200,
        );
      });

      final data = await httpClient(
        method: 'GET',
        baseUrl: 'https://api.example.com',
        endpoint: 'todos/1',
        httpClient: mock,
      );

      expect(data, {'id': 1, 'title': 'Test'});
    });

    test('re-throws HttpClientException', () async {
      final mock = MockClient((_) async {
        return http.Response(
          jsonEncode({'error': 'Not Found'}),
          404,
        );
      });

      expect(
        () => httpClient(
          method: 'GET',
          baseUrl: 'https://api.example.com',
          endpoint: 'todos/999',
          httpClient: mock,
        ),
        throwsA(isA<HttpClientException>()),
      );
    });

    test('wraps connection errors in Exception with errorMessage prefix', () async {
      final mock = MockClient((_) async {
        throw Exception('Connection refused');
      });

      expect(
        () => httpClient(
          method: 'GET',
          baseUrl: 'https://api.example.com',
          endpoint: 'todos/1',
          errorMessage: 'Request failed',
          httpClient: mock,
        ),
        throwsA(
          isA<Exception>().having(
            (e) => e.toString(),
            'message',
            allOf(
              contains('Request failed'),
              contains('[Connection error]'),
            ),
          ),
        ),
      );
    });

    test('closes client even on error (finally block)', () async {
      var requestCount = 0;
      final mock = MockClient((_) async {
        requestCount++;
        return http.Response('', 500);
      });

      // The function should throw, but the client should still be cleaned up
      try {
        await httpClient(
          method: 'GET',
          baseUrl: 'https://api.example.com',
          endpoint: 'todos/1',
          httpClient: mock,
        );
      } on HttpClientException {
        // Expected
      }

      // Verify the request was actually made (client was used then closed)
      expect(requestCount, 1);
    });
  });
}

File: code/ts/test/http/http_client.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { httpClient } from '../../src/http/http_client';
import { HttpClientException } from '../../src/http/http_client_exception';

function mockResponse(body: string, status: number): Response {
  return {
    status,
    headers: new Headers(),
    text: () => Promise.resolve(body),
  } as unknown as Response;
}

describe('httpClient()', () => {
  beforeEach(() => {
    vi.stubGlobal('fetch', vi.fn());
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it('returns parsed data on success', async () => {
    vi.mocked(fetch).mockResolvedValue(
      mockResponse(JSON.stringify({ id: 1, title: 'Test' }), 200),
    );

    const data = await httpClient({
      method: 'GET',
      baseUrl: 'https://api.example.com',
      endpoint: 'todos/1',
    });

    expect(data).toEqual({ id: 1, title: 'Test' });
  });

  it('re-throws HttpClientException', async () => {
    vi.mocked(fetch).mockResolvedValue(
      mockResponse(JSON.stringify({ error: 'Not Found' }), 404),
    );

    await expect(
      httpClient({
        method: 'GET',
        baseUrl: 'https://api.example.com',
        endpoint: 'todos/999',
      }),
    ).rejects.toThrow(HttpClientException);
  });

  it('wraps connection errors in Error with errorMessage prefix', async () => {
    vi.mocked(fetch).mockRejectedValue(new TypeError('fetch failed'));

    await expect(
      httpClient({
        method: 'GET',
        baseUrl: 'https://api.example.com',
        endpoint: 'todos/1',
        errorMessage: 'Request failed',
      }),
    ).rejects.toThrow(/Request failed.*\[Connection error\]/);
  });

  it('closes client even on error (finally block)', async () => {
    let callCount = 0;
    vi.mocked(fetch).mockImplementation(async () => {
      callCount++;
      return mockResponse('', 500);
    });

    try {
      await httpClient({
        method: 'GET',
        baseUrl: 'https://api.example.com',
        endpoint: 'todos/1',
      });
    } catch {
      // Expected
    }

    expect(callCount).toBe(1);
  });
});
